File: src/praisonai-mobile/adapters/src/tauri/secrets.ts
/**
 * SecretsPort over the platform keychain in `src-tauri/plugins/secrets`.
 *
 * WHY THIS EXISTS. `app/src/platform.ts` branched `storage` per platform and
 * handed `createWebSecrets()` -- a module-scoped `Map` -- to BOTH. So on a
 * phone the API key worked for exactly as long as the process lived, and the
 * user re-entered it on every single launch. `core/src/ports/secrets.ts` opens
 * by saying the port IS "the keychain (iOS) and keystore (Android)"; this is
 * the adapter that makes the sentence true on a device.
 *
 * `isHardwareBacked` IS TRUE HERE, and that is a claim, not an assumption.
 * `detectPlatform` is synchronous -- `ShellPort.insets` has to be readable
 * during the first paint -- so this file cannot ask the host "are you iOS?"
 * before answering. The invariant is kept one layer down instead: the Rust
 * plugin refuses every call on a platform with no hardware store rather than
 * falling back to a file, so a secret that was stored AT ALL through this
 * adapter is in the Keychain or the Keystore. The alternative -- a flag that
 * starts false and flips when a probe lands -- would show the "not protected"
 * warning on a keychain-backed phone for the first frames of the settings
 * screen, which is a warning that means nothing the moment it is wrong once.
 *
 * `invokeStrict`, not `invoke`. The forgiving `invoke` resolves to `null` on
 * failure, and `null` is `get`'s word for "no such secret" -- so on the
 * forgiving one a failing keychain would present as a user who never
 * configured a key, and the app would ask for it again instead of saying it
 * could not be read.
 */
import type { SecretRef, SecretsPort } from "../../../core/src/ports/secrets.ts";
import type { StrictInvoke } from "./storage.ts";

/**
 * The seam with `src-tauri/src/secrets.rs`, which declares the same four
 * strings as `pub const`s. `tools/secrets-seam.test.mjs` reads both files and
 * compares them: a rename on one side alone is silent, and silent here means
 * the settings screen says "Not set" forever while the key sits in the
 * keychain -- and every attempt to fix it by re-typing the key also fails.
 */
export const SECRET_COMMANDS = {
  read: "secret_read",
  write: "secret_write",
  remove: "secret_remove",
  has: "secret_has",
} as const;

export interface TauriSecretsDeps {
  readonly invoke: StrictInvoke;
}

/**
 * A reply that is not the shape the command promises is an I/O FAILURE, not an
 * absence -- the same distinction `createTauriStorage` draws, for the same
 * reason. A native `undefined` read as `null` would report a broken keychain
 * as "no key configured", and the user would paste a working key into a store
 * that cannot keep it.
 */
function asStringOrNull(command: string, value: unknown): string | null {
  // `undefined` as well as `null`, unlike `createTauriStorage`. A Rust
  // `Ok(None)` reaches the webview as JSON `null`; `undefined` is accepted
  // alongside it because the two are the same absence to every caller of this
  // port, and because nothing here depends on telling them apart -- an I/O
  // failure REJECTS rather than returning either, which is the distinction
  // that actually matters and the reason `invokeStrict` is used. Exercised on
  // a device: a fresh install with no key stored reports the SDK's "the
  // OPENAI_API_KEY environment variable is missing", not a shape error.
  if (value === null || value === undefined) return null;
  if (typeof value === "string") return value;
  throw new Error(`${command} returned ${typeof value}, not a string or null`);
}

function asBoolean(command: string, value: unknown): boolean {
  if (typeof value !== "boolean") {
    throw new Error(`${command} returned ${typeof value}, not a boolean`);
  }
  return value;
}

export function createTauriSecrets(deps: TauriSecretsDeps): SecretsPort {
  // The argument names are the Rust parameter names. Tauri maps a JS key onto
  // a `snake_case` Rust parameter; both of these are a single lowercase word
  // precisely so there is no case convention to get wrong -- a mismatch is a
  // deserialisation error on EVERY call, reported to the webview as "could not
  // save" with nothing pointing at the cause.
  const of = (ref: SecretRef): Record<string, unknown> => ({
    slot: ref.slot,
    account: ref.account,
  });

  return {
    // See the header. Guaranteed by the plugin refusing rather than degrading.
    isHardwareBacked: true,

    /**
     * Presence, through its OWN command.
     *
     * Rule 2 of the port: `has()` "exists so the UI can render 'configured'
     * without reading the value" and "must not fault the value into memory".
     * `get(ref) !== null` would satisfy every behavioural test in the contract
     * and copy the user's API key out of the keychain into the webview's heap
     * on every repaint of the settings screen -- where a content-script bug, a
     * crash dump or a devtools attach can reach it. So the check goes to Rust
     * as a presence query and comes back as a boolean.
     */
    async has(ref) {
      const present = await deps.invoke(SECRET_COMMANDS.has, of(ref));
      return asBoolean(SECRET_COMMANDS.has, present);
    },

    async get(ref) {
      const value = await deps.invoke(SECRET_COMMANDS.read, of(ref));
      return asStringOrNull(SECRET_COMMANDS.read, value);
    },

    async set(ref, value) {
      await deps.invoke(SECRET_COMMANDS.write, { ...of(ref), value });
    },

    async delete(ref) {
      await deps.invoke(SECRET_COMMANDS.remove, of(ref));
    },
  };
}

File: src/praisonai-mobile/src-tauri/src/secrets.rs
//! The four commands the webview reaches the platform secret store through.
//!
//! The store itself is `plugins/secrets` -- the Keychain on Apple, the
//! Keystore-backed `EncryptedSharedPreferences` on Android, and a refusal
//! anywhere else. This module is the boundary in front of it, and it exists for
//! one reason the store cannot supply: **the slot is a closed union**.
//!
//! `core/src/ports/secrets.ts` rule 3 spells out why. `SecretSlot` is five
//! names, not a string, because "a free-form name lets a bug write an
//! attacker-influenced string into the keychain service namespace" -- and the
//! slot becomes the keychain SERVICE (`service_for` in the plugin) and part of
//! the Android preferences key. TypeScript's union is erased at runtime; the
//! webview is where a compromised page would be; so the union has to exist a
//! second time, in Rust, as an allowlist that cannot be talked around. Exactly
//! the argument `store.rs`'s `NAMESPACES` makes for a directory name, one
//! sensitive namespace over.
//!
//! The command names are duplicated in `adapters/src/tauri/secrets.ts` and
//! compared by `tools/secrets-seam.test.mjs`: a rename on one side alone is
//! silent, and silent here means the settings screen reports "Not set" forever
//! while the user's key sits in the keychain.

use tauri::{AppHandle, Runtime};
use tauri_plugin_secrets::SecretsExt;

/// The command names the webview invokes. `adapters/src/tauri/secrets.ts`
/// names the same four strings.
pub const CMD_READ: &str = "secret_read";
pub const CMD_WRITE: &str = "secret_write";
pub const CMD_REMOVE: &str = "secret_remove";
pub const CMD_HAS: &str = "secret_has";

/// The slots `core/src/ports/secrets.ts` declares, and no others.
///
/// Not a nicety: the slot is the keychain service name. An arbitrary string
/// here would let the webview name a service belonging to something else, and
/// on Android would let it write outside the entry namespace this app owns.
pub const SLOTS: [&str; 5] = ["openai", "anthropic", "google", "openrouter", "custom"];

/// The longest account name that will be stored.
///
/// Refused with a NAME rather than handed to the platform, which answers a
/// too-long keychain attribute with `errSecParam` -- a number that tells the
/// user nothing and points at no cause.
pub const MAX_ACCOUNT_LEN: usize = 200;

/// A slot, or an error naming the one that was refused.
pub fn checked_slot(slot: &str) -> Result<&str, String> {
    SLOTS
        .iter()
        .find(|known| **known == slot)
        .copied()
        .ok_or_else(|| format!("unknown secret slot {slot:?}"))
}

/// An account name, or an error saying why not.
pub fn checked_account(account: &str) -> Result<&str, String> {
    if account.is_empty() {
        return Err("a secret account may not be empty".to_string());
    }
    if account.len() > MAX_ACCOUNT_LEN {
        return Err(format!(
            "secret account is too long ({} bytes, limit {MAX_ACCOUNT_LEN})",
            account.len()
        ));
    }
    Ok(account)
}

/// Both checks, in the one order every command uses them.
fn checked(slot: &str, account: &str) -> Result<(), String> {
    checked_slot(slot)?;
    checked_account(account)?;
    Ok(())
}

// ---------------------------------------------------------------------------
// the Tauri commands
// ---------------------------------------------------------------------------
//
// Every argument is a single lowercase word on purpose, exactly as in
// `store.rs`: Tauri 2 maps JS `camelCase` onto Rust `snake_case` parameters,
// and a two-word name is the standing way to get a command that rejects every
// call with a deserialisation error the webview reports as "could not save".

/// `Ok(None)` for a secret that was never set. Absence is not an error.
#[tauri::command]
pub fn secret_read<R: Runtime>(
    app: AppHandle<R>,
    slot: String,
    account: String,
) -> Result<Option<String>, String> {
    checked(&slot, &account)?;
    app.secrets()
        .read(&slot, &account)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn secret_write<R: Runtime>(
    app: AppHandle<R>,
    slot: String,
    account: String,
    value: String,
) -> Result<(), String> {
    checked(&slot, &account)?;
    app.secrets()
        .write(&slot, &account, &value)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn secret_remove<R: Runtime>(
    app: AppHandle<R>,
    slot: String,
    account: String,
) -> Result<(), String> {
    checked(&slot, &account)?;
    app.secrets()
        .remove(&slot, &account)
        .map_err(|error| error.to_string())
}

/// Presence only. See rule 2 of `core/src/ports/secrets.ts`: this must not
/// bring the value into memory, which is why it is a command of its own rather
/// than `secret_read(..).is_some()` in the adapter.
#[tauri::command]
pub fn secret_has<R: Runtime>(
    app: AppHandle<R>,
    slot: String,
    account: String,
) -> Result<bool, String> {
    checked(&slot, &account)?;
    app.secrets()
        .has(&slot, &account)
        .map_err(|error| error.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_command_names_are_the_ones_the_webview_invokes() {
        // Literals on purpose, exactly as store.rs does for its five. A rename
        // here is silent on a device: every secret call rejects and the app
        // keeps running, reporting "Not set" for a key that is stored.
        assert_eq!(CMD_READ, "secret_read");
        assert_eq!(CMD_WRITE, "secret_write");
        assert_eq!(CMD_REMOVE, "secret_remove");
        assert_eq!(CMD_HAS, "secret_has");
    }

    #[test]
    fn the_slots_are_the_five_the_port_declares() {
        assert_eq!(
            SLOTS,
            ["openai", "anthropic", "google", "openrouter", "custom"]
        );
    }

    #[test]
    fn a_slot_outside_the_union_is_refused_rather_than_named() {
        // Rule 3. Each of these, accepted, becomes a keychain service name or
        // an Android preferences namespace of the caller's choosing.
        for bad in [
            "",
            "OpenAI",
            "openai ",
            "../../etc",
            "ai.praison.mobile.secrets.openai",
            "com.apple.account",
            "openai:extra",
        ] {
            let refused = checked_slot(bad);
            assert!(refused.is_err(), "{bad:?} was accepted as a slot");
            assert!(
                refused.unwrap_err().contains("unknown secret slot"),
                "{bad:?} was refused without saying why"
            );
        }
    }

    #[test]
    fn every_declared_slot_is_accepted() {
        // The control. A checker that refused everything would satisfy the case
        // above while making the app unable to store anything at all.
        for slot in SLOTS {
            assert_eq!(checked_slot(slot).unwrap(), slot);
        }
    }

    #[test]
    fn no_slot_contains_the_separator_the_android_key_is_built_with() {
        // `SecretsPlugin.kt` composes its preferences key as "$slot:$account".
        // That is injective ONLY while no slot contains a colon: with one,
        // slot "a:b" + account "c" and slot "a" + account "b:c" would be a
        // single entry, and adding one profile's key would overwrite another's.
        for slot in SLOTS {
            assert!(!slot.contains(':'), "{slot:?} would collide two entries");
        }
    }

    #[test]
    fn an_empty_or_over_long_account_is_refused_by_name() {
        assert!(checked_account("").unwrap_err().contains("may not be empty"));
        let long = "a".repeat(MAX_ACCOUNT_LEN + 1);
        assert!(checked_account(&long).unwrap_err().contains("too long"));
        assert_eq!(checked_account("default").unwrap(), "default");
        assert_eq!(
            checked_account(&"a".repeat(MAX_ACCOUNT_LEN)).unwrap(),
            "a".repeat(MAX_ACCOUNT_LEN)
        );
    }

    #[test]
    fn the_pair_is_checked_before_anything_reaches_the_platform() {
        // Both halves, in one call, because a command that checked only the
        // slot would still hand an unbounded account to the keychain.
        assert!(checked("openai", "default").is_ok());
        assert!(checked("nope", "default").is_err());
        assert!(checked("openai", "").is_err());
    }
}
